barretenberg/cpp/src/barretenberg/vm2/simulation/execution.cpp

```diff
@@ -0,0 +1,180 @@
+#include "barretenberg/vm2/simulation/execution.hpp"
+
+#include <algorithm>
+#include <concepts>
+#include <cstdint>
+#include <functional>
+
+#include "barretenberg/vm2/common/instruction_spec.hpp"
+#include "barretenberg/vm2/common/memory_types.hpp"
+#include "barretenberg/vm2/common/opcodes.hpp"
+#include "barretenberg/vm2/simulation/addressing.hpp"
+#include "barretenberg/vm2/simulation/context.hpp"
+#include "barretenberg/vm2/simulation/events/execution_event.hpp"
+
+namespace bb::avm2::simulation {
+
+void Execution::add(ContextInterface& context, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr)
+{
+    alu.add(context, a_addr, b_addr, dst_addr);
+}
+
+// TODO: My dispatch system makes me ahve a uint8_t tag. Rethink.
+void Execution::set(ContextInterface& context, MemoryAddress dst_addr, uint8_t tag, MemoryValue value)
+{
+    context.get_memory().set(dst_addr, std::move(value), static_cast<MemoryTag>(tag));
+}
+
+void Execution::mov(ContextInterface& context, MemoryAddress src_addr, MemoryAddress dst_addr)
+{
+    auto& memory = context.get_memory();
+    auto [value, tag] = memory.get(src_addr);
+    memory.set(dst_addr, std::move(value), tag);
+}
+
+// TODO: This will need to happen in its own gadget in any case.
+void Execution::call(ContextInterface& context, MemoryAddress addr)
+{
+    auto& memory = context.get_memory();
+
+    const auto [contract_address, _] = memory.get(addr);
+    std::vector<FF> calldata = {};
+
+    auto nested_context = context_provider.make(std::move(contract_address),
```

Comment:
We can still choose to make a recursive call to `execute` here. Depends on how we set up things but I suspect that will be the case. We will likely still need a context stack though, we need to manage it explicitly so that we have the events we need.